### PR TITLE
Cool to hic

### PR DESCRIPTION
**Pull request by fubar2**
Will this do for a start @bgruening 
Sorry it took so long - I was testing a bad cool file. Once I found a valid one, it worked fine.
uses hictk
needs logic to consider case of mcool 
hictk cl expects things to have informative extensions.
must take a look at the python bindings - probably more convenient..
Output works with JBrowse2

image

### CODE REVIEW

File: tools/cool_to_hic/.shed.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+name: cool_to_hic
+owner: fubar2
```
Comment (bgruening) [inline tools/cool_to_hic/.shed.yml:2]: bgruening

File: tools/cool_to_hic/cool_to_hic.xml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  <!--Created by toolfactory@galaxy.org at 01/02/2024 07:02:54 using the Galaxy Tool Factory.-->
+  <description>Convert .cool to .hic binary for JBrowse viewing or other use</description>
+  <requirements>
+    <requirement version="0.0.8" type="package">hictk</requirement>
+  </requirements>
+  <version_command><![CDATA[echo "0.0.8"]]></version_command>
+  <command><![CDATA[bash '$runme']]></command>
+  <configfiles>
+    <configfile name="runme"><![CDATA[ln -s $incool in.$incool.ext
+hictk convert in.$incool.ext out.hic
```
Comment (bgruening) [inline tools/cool_to_hic/cool_to_hic.xml:12]: Single quotes are missing

File: tools/cool_to_hic/cool_to_hic.xml [hunk 3]
```diff
@@ -0,0 +1,75 @@
+<tool name="cool_to_hic" id="cool_to_hic" version="0.0.8" profile="22.05">
+  <!--Source in git at: https://github.com/fubar2/galaxy_tf_overlay-->
+  <!--Created by toolfactory@galaxy.org at 01/02/2024 07:02:54 using the Galaxy Tool Factory.-->
+  <description>Convert .cool to .hic binary for JBrowse viewing or other use</description>
```
Comment (bgruening) [inline tools/cool_to_hic/cool_to_hic.xml:4]: Maybe remove jbrowse from the description? It's a normal generic converter.

File: tools/cool_to_hic/cool_to_hic.xml [hunk 4]
```diff
@@ -0,0 +1,75 @@
+<tool name="cool_to_hic" id="cool_to_hic" version="0.0.8" profile="22.05">
```
Comment (bgruening) [inline tools/cool_to_hic/cool_to_hic.xml:1]: Maybe the id as hictk? So we can extend it later for more conversions?

File: tools/cool_to_hic/.shed.yml (hunk 1 above)
Comment (fubar2) [reply tools/cool_to_hic/.shed.yml:2]: done

File: tools/cool_to_hic/cool_to_hic.xml (hunk 3 above)
Comment (fubar2) [reply tools/cool_to_hic/cool_to_hic.xml:4]: done

File: tools/cool_to_hic/cool_to_hic.xml (hunk 4 above)
Comment (fubar2) [reply tools/cool_to_hic/cool_to_hic.xml:1]: done.
name and id are identical for toolfactory tools

File: tools/cool_to_hic/cool_to_hic.xml (hunk 2 above)
Comment (fubar2) [reply tools/cool_to_hic/cool_to_hic.xml]: fixed

**Comment by fubar2** [comment]
Yay - the matrix.mcool and cool samples in lib/datatypes/test convert to hic without hictk complaining. Not viewed as don't know the genome but at least it seems to work. Need a scool file to try...

**Comment by fubar2** [comment]
You will also need a hic binary:Binary subclass datatype added to datatypes_conf.xml

**Comment by fubar2** [comment]
Having problems with hicexplorer cool but not with the ones in lib/galaxy/datatypes/test

Testing on the command line to remove all other variables, cool samples like the ones in lib/galaxy/datatypes/test convert with hictk. 

When I make a matrix using the hicexplorer dm4 tutorial and convert to cool with the hicexplorer converter, it will not convert with hictk.  

`hictk validate` says it's a valid cooler so I dunno.... 
Could be a hictk problem for all I know if hicexplorer cool is widely compatible elsewhere?

There is something fishy going on - not sure I broke anything new, but it looks like hictk is not compatible with hicexplorer cool which means it's useless for galaxy.

**Comment by fubar2** [comment]
It's possible to open both the hicexplorer sample with the python api!!
`
import hictkpy as htk

Invoked with: 'matrix.cool'; kwargs: resolution=None
>>> f = htk.File("matrix.cool", resolution=1)
>>> f.attributes()
{'bin_size': 5000, 'bin_type': 'fixed', 'format': 'HDF5::Cooler', 'format_version': 2, 'storage-mode': None, 'creation-date': '2017-11-15T19:04:25.627510', 'generated-by': 'cooler-0.7.6', 'assembly': 'unknown', 'metadata': '{}', 'format-url': 'https://github.com/mirnylab/cooler', 'nbins': 33754, 'nchroms': 15, 'nnz': 35857, 'sum': 37321, 'cis': None}
>>> f = htk.File("hicex_dm4.cool", resolution=1)
>>> f.attributes()
{'bin_size': 10000, 'bin_type': 'fixed', 'format': 'HDF5::Cooler', 'format_version': 3, 'storage-mode': 'symmetric-upper', 'creation-date': '2024-02-03T03:46:02.531452', 'generated-by': 'HiCMatrix-16', 'assembly': "b'dm3'", 'metadata': '{"format": "HDF5::Cooler", "format-url": "https://github.com/mirnylab/cooler", "generated-by": "HiCMatrix-16", "generated-by-cooler-lib": "cooler-0.9.3", "tool-url": "https://github.com/deeptools/HiCMatrix", "matrix-generated-by": "b\'HiCExplorer-3.7.2\'", "matrix-generated-by-url": "b\'https://github.com/deeptools/HiCExplorer\'", "genome-assembly": "b\'dm3\'"}', 'format-url': 'https://github.com/mirnylab/cooler', 'nbins': 50717, 'nchroms': 37688, 'nnz': 19482, 'sum': 4314105, 'cis': None}
>>>

**Comment by fubar2** [comment]
Got a version of jbrowse2 that accepts hicexplorer cool files - here's one from following the GTN tutorial.
Note the name of the track file - a cool output from the matrix builder. 
New conda container to add hictk so let me know when suits for manual update but will be very `cool`

image

**Comment by bgruening** [comment]
I need to add a datatype before, otherwise I would merge it already.

**Comment by bgruening** [comment]
We call it `juicebox_hic` thanks @lldelisle for the suggestion. https://github.com/aidenlab/Juicebox/blob/master/HiCFormatV8.md

**Review by bgruening — APPROVED** [review]

**Review by fubar2 — COMMENTED** [review]
All good for me.

**Comment by fubar2** [comment]
@bgruening: 
What's the official new entry for datatypes_conf.xml for this tool?
If no longer 'hic', need to adjust JBrowse2 XML...

**Pull request merged**